freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/model/UsersStateDifference.java

```diff
@@ -78,28 +81,45 @@ public static UsersStateDifference fromUmsAndIpaUsersStates(UsersState umsState,
 
         Multimap<String, String> groupMembershipToAdd = HashMultimap.create();
         umsGroupMembership.forEach((group, user) -> {
-            LOGGER.info("Evaluation group = {} and user = {}", group, user);
             if (!ipaGroupMembership.containsEntry(group, user)) {
-                LOGGER.info("adding");
+                LOGGER.debug("adding user : {} to group : {}", user, group);
                 groupMembershipToAdd.put(group, user);
             }
         });
+        LOGGER.info("groupMembershipToAdd size= {}", groupMembershipToAdd.size());
 
         Multimap<String, String> groupMembershipToRemove = HashMultimap.create();
         ipaGroupMembership.forEach((group, user) -> {
-            LOGGER.info("Evaluation group = {} and user = {}", group, user);
             if (!umsGroupMembership.containsEntry(group, user)) {
-                LOGGER.info("removing");
+                LOGGER.debug("removing user : {} to group : {}", user, group);
                 groupMembershipToRemove.put(group, user);
             }
         });
 
+        LOGGER.info("groupMembershipToRemove size= {}", groupMembershipToRemove.size());
+
+        Set<FmsUser> usersToRemove =
+            getUsersToBeRemoved(umsState.getGroupMembership().get(UserServiceConstants.USERSYNC_INTERNAL_GROUP),
+                                ipaState.getUsers());
+        LOGGER.info("usersToRemove size= {}", usersToRemove.size());
+
         return new UsersStateDifference(
-                Set.copyOf(Sets.difference(umsState.getGroups(), ipaState.getGroups())),
-                Set.copyOf(Sets.difference(umsState.getUsers(), ipaState.getUsers())),
-                Set.copyOf(Sets.difference(ipaState.getGroups(), umsState.getGroups())),
-                Set.copyOf(Sets.difference(ipaState.getUsers(), umsState.getUsers())),
-                groupMembershipToAdd,
-                groupMembershipToRemove);
+            Set.copyOf(Sets.difference(umsState.getGroups(), ipaState.getGroups())),
+            Set.copyOf(Sets.difference(umsState.getUsers(), ipaState.getUsers())),
+            Set.copyOf(Sets.difference(ipaState.getGroups(), umsState.getGroups())),
+            usersToRemove,
+            groupMembershipToAdd,
+            groupMembershipToRemove);
```

Comment:
I think we need to also consider `USERSYNC_INTERNAL_GROUP` when calculating the groupMembershipToRemove so that we don't remove group membership for users that are not managed by FMS.

Reply:
If any group created outside of fms, we dont support that right now. If there is a requirement like that then we need to know upfront which group needs to be  taken care.